File: LeetCode/Medium/0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        length = 0
        curr = head

        while curr:
            length += 1
            curr = curr.next

        part_size = length // k
        extra_nodes = length % k

        curr = head
        ret = []

        for i in range(k):
            part_head = curr
            current_part_size = part_size + (1 if extra_nodes > 0 else 0)
            extra_nodes -= 1
            
            for j in range(current_part_size - 1):
                if curr:
                    curr = curr.next
            
            if curr:
                next_part_head = curr.next
                curr.next = None
                curr = next_part_head
            
            ret.append(part_head)
        
        return ret

File: LeetCode/Medium/0725-split-linked-list-in-parts/0725-split-linked-list-in-parts.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        int n = 0;
        ListNode* curr = head;
        
        while (curr){
            ++n;
            curr = curr->next;
        }
        
        curr = head;
        
        vector<ListNode*> ret(k);
        
        int need = n / k;
        int remain = n % k;
        
        for (int i = 0; i < k; ++i){
            ListNode* temp = curr;
            int count = need;
            
            if (remain > 0) {
                remain--;
                ++count;
            }
            
            for (int j = 0; j < count - 1; ++j){
                if (!curr) break;
                curr = curr->next;
            }
            
            if (curr){
                ListNode* next = curr->next;
                curr->next = nullptr;
                curr = next;
            }

            ret[i] = temp;
        }
        
        return ret;
    }
};